File: yandex/yandex_43412_ml/c_cpp.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1000000007;

int inv(int a) {
    int r = 1, t = a, k = MOD - 2;
    while (k) {
        if (k & 1) r = (long long) r * t % MOD;
        t = (long long) t * t % MOD;
        k >>= 1;
    }
    return r;
}

int main() {
    int n, k;
    cin >> n >> k;

    vector<int> p(n, k);

    int a = 1, b = 0;
    for (int i = 0; i < n; i++) {
        a = (long long) a * inv(p[i]) % MOD;
        b = (long long) b * inv(p[i]) % MOD;
        a = (a + (long long) (p[i] - 1) * inv(p[i])) % MOD;
        b = (b - inv(p[i]) + MOD) % MOD;
    }
    cout << (long long) (((MOD - b) * inv(a) % MOD) / (k + 1) * k) << "\n";
    return 0;
}

File: yandex/yandex_43412_ml/c_cpp.py
MOD = 1000000007

def inv(a):
    r = 1
    t = a
    k = MOD - 2
    while k:
        if k & 1:
            r = (r * t) % MOD
        t = (t * t) % MOD
        k >>= 1
    return r

n, k = map(int, input().split())
p = inv(k) % MOD
a = 1
b = 0

for i in range(n):
    a = (a * inv(p)) % MOD
    b = (b * inv(p)) % MOD
    a = (a + (p - 1) * inv(p)) % MOD
    # this should be
    b = (b + inv(p) - 1) % MOD

result = (MOD - b) * inv(a) % MOD
print(result)